module/zfs/zfs_log.c

```diff
@@ -257,7 +257,8 @@ zfs_xattr_owner_unlinked(znode_t *zp)
 	if (tzp != zp)
 		zrele(tzp);
 #else
-	zhold(zp);
+	/* Shouldn't be first ref, so zhold() cannot return NULL */
+	VERIFY3P(zhold(zp), !=, NULL);
```

Comment:
So on FreeBSD `zhold()` maps to `vhold()` with returns a `(void)` type.  So we'll need to confine these checks to the Linux specific bits.  It might be best to call `igrab()` directly in the Linux code.

Reply:
Good point, although since zhold still may or may not return NULL in the common code, the common code should go through the motions of checking the return even if the build platform is FreeBSD, right?  Or utterly enforce a void-return everywhere for consistency, and *only* check a bare `igrab()` and VERIFY() zhold's wrapping of igrab() - is that better?

Reply:
>  Or utterly enforce a void-return everywhere for consistency, and only check a bare igrab() and VERIFY() zhold's wrapping of igrab() - is that better?

My feeling is this would be best way to go.  We could even define `zhold()` on Linux as something like this to make it clear it can only be called when we already have a reference.  This is the behavior we want in the common code since that's a requirement for FreeBSD, hence why it can't fail.  Then as you said, we explicitly use `igrab()` everywhere in the Linux code where it's possible for it to fail.

```c
/* Must already be referenced so igrab() cannot return NULL */
#define zhold(zp) VERIFY3P(igrab(ZTOI((zp))), !=, NULL)
```